gym_wrapper/base_gym_env.py

```diff
@@ -26,6 +27,8 @@ def __init__(
         Arguments:
             level (str): path to the config file to load. Most settings should be set by this config file.
             frame_skip (int): how many frames should be advanced per action. 1 = take action on every frame. Default: 1.
+            max_buttons_pressed (int): how many binary buttons can be pressed at once. 1 results in a discrete action space.
+                                       Otherwise results in a MultiDiscrete action space. Default: 1
```

Comment:
Hmm the point of this is that you limit the number of Discrete choices: if you do all possible combinations of N buttons, you get a lot of options. Limiting to "M pressed buttons at once" limits the Discrete space size.

MultiDiscrete bypasses this issue completely, and I do not think it should be limited in this way. I would maybe change this to `if max_buttons_pressed is None, then use MultiDiscrete space instead`. This also allows people to use Discrete space with more than one button pressed at once in case they do not have an implementation for MultiDiscrete space.

Clarification: in `MultiDiscrete` space, the idea is that each button has on/off or 0/1 states, and you individually choose the on/off state for each button.

Reply:
Ah I guess I misunderstood how `MultiDiscrete` works. Here was my thought process:

Say there are 4 in-game buttons defined (`ATTACK`, `MOVE_LEFT`, `MOVE_RIGHT`, `MOVE_FORWARD`), and `max_buttons_pressed=2`. All possible combinations of N buttons is 
```
[0, 0, 0, 0] // no op
[1, 0, 0, 0] // ATTACK
[0, 1, 0, 0] // MOVE_LEFT
[0, 0, 1, 0] // MOVE_RIGHT
[0, 0, 0, 1] // MOVE_FORWARD
[1, 1, 0, 0] // ATTACK + MOVE_LEFT
[1, 0, 1, 0] // ATTACK + MOVE_RIGHT
[1, 0, 0, 1] // ATTACK + MOVE_FORWARD
[0, 1, 1, 0] // MOVE_LEFT + MOVE_RIGHT (effectively no_op...)
[0, 1, 0, 1] // MOVE_LEFT + MOVE_FORWARD
[0, 0, 1, 1] // MOVE_RIGHT + MOVE_FORWARD
```
To avoid the action space becoming `Discrete(11)`, which with a more complicated example becomes way too many options for all possible combinations, I thought the users action would instead be a list with $(0 <$`len(action)`$<= 2)$ (could can also allow for just an `int` easily enough) where each value is within $[0, 3]$. An example action could be `[1, 2]` which corresponds to `[0, 1, 1, 0]`. I figured this was `MultiDiscrete([2, 4])`, but I misunderstood how `MultiDiscrete` works. 

Is this more akin to just having the action space be `Discrete(4)` and allowing for a list of ints (or just a single int)? How were you thinking this action space would work? Also are we forcing the user to provide `max_buttons_pressed` number of actions, or just up to `max_buttons_pressed` number of actions.

Reply:
Ah, fair confusion: `MultiDiscrete` is essentially multiple `Discrete` choices.

If we only had a single button we can press down, then we would have `Discrete(2)` space (two choices) or an equal `MultiDiscrete([2,])` space.

If we have two buttons, both can be pressed down or be up, so both have `Discrete(2)` space. We _could_ do all possible combinations and create the `Discrete(2 * 2)` space, but we can also do `MultiDiscrete([2,2])` space. Essentially, you choose for each button individually if button should be down or not.

For four buttons, this would be `MultiDiscrete([2,2,2,2])`.

I was thinking that if `max_buttons_pressed=None` (or -1 or something special), then you could just represent all binary buttons with `MultiDiscrete`

Reply:
I see.

Final clarification then: 

For the case where `max_buttons_pressed` is positive, would you like the action space to be a `gym.spaces.Tuple` of `Discrete(2)` with `len == max_buttons_pressed`? (And obviously if `max_buttons_pressed==1` we can rid the Tuple, and just have `Discrete(2)`)?

In this case, is the user still sending in a list of indices indicating which binary buttons are "pressed", i.e. `[0, 2]` would create the environment action: `[1, 0, 1, 0]`? (and again they can easily just send an int for a single action, etc, that's no issue)

Reply:
To be precise, something like this:

if `max_buttons_pressed > 0`: binary action space is `Tuple` of `Discrete(2)` with `len == max_buttons_pressed`.
(potentially can have a special case when `max_buttons_pressed == 1`, binary action space is `Discrete(2)`)

if `max_buttons_pressed < 0`: binary action space is `MultiDiscrete([2] * -max_buttons_pressed) `.

Reply:
Ah, sorry, this is what I had in mind (but I can see where the confusions are coming up; this suggestion is not elegant either 😅 ):

If `max_buttons_pressed > 0`: Create a _single_ `Discrete(n)` space for pressing the binary buttons, where `n = 2 * 2 * 2... = 2^num_buttons` (i.e. all possible combinations of pressing `num_buttons` different buttons).

If `max_buttons_pressed == -1 (or 0)`: Create a single `MultiDiscrete([2, 2, 2, ...]) = MultiDiscrete([2,] * num_buttons)` space, where the press action is decided for each button separately.

Any other `max_buttons_pressed` raises an error which tells of these two options.